### ISSUE TITLE
508-defect-1 ❗ Launchblocker [SCREENREADER, KEYBOARD, ZOOM]: The collapse/expand implementation MUST not trap users

### ISSUE THREAD

**Issue by jenstrickland** [department-of-veterans-affairs/va.gov-team#10997]
# [508-defect-1 :exclamation: Launchblocker](https://github.com/department-of-veterans-affairs/va.gov-team/blob/master/platform/accessibility/guidance/defect-severity-rubric.md#508-defect-1)

**Feedback framework**
- **❗️ Must** for if the feedback must be applied 
- **⚠️Should** if the feedback is best practice 
- **✔️ Consider** for suggestions/enhancements 

## Description

The current sidenav uses a method of a fixed `height: 45px` parent container with `overflow:hidden`. This creates a number of issues, and traps a screen reader user.

- The fixed height parent container is used to hide the sidenav items, but they are still available by tab or reading the page. If a user tabs through interactive items, they become stuck in the sidenav children, with the nav not expanded. Figuring out how to get out is confusing.
- As it is now, if the screen reader reads through the sidenav past the "In this section" the user can get stuck on the children list items.
- Additionally, if a user finds a way to increase their font-size, the text will exceed the fixed 45px height. 
 
The collapse/expand implementation **must** be replaced with the `<ul>` using `display:none` in the collapsed state. See the [Accordions documentation in the VA.gov design system](https://design.va.gov/components/accordions).

## Point of Contact

**VFS Point of Contact:** Jennifer

## Acceptance Criteria

- As a keyboard user, I want to navigate the sidenav without getting trapped.
- As a screen reader user, I want to navigate the sidenav without getting trapped.

## Environment

* Operating System: all
* Browser: all
* Screenreading device: all
* Server destination: production

## Steps to Recreate

1. Enter https://www.va.gov/pittsburgh-health-care/locations/pittsburgh-va-medical-center-university-drive/ in browser
2. Tab through the sidenav items — without expanding it
3. Verify that is possible, and that it is not possible to expand the sidenav once the focus is in the children items
4. Start screen reader
5. Reload the page and read the page
6. Verify that the screen reader reads the children when the sidenav is in the collapsed state
7. Verify that it is not possible to expand the sidenav when focus is in the children

## Possible Fixes (optional)

Use the [Accordions documentation in the VA.gov design system](https://design.va.gov/components/accordions) to use `display:none' on the `ul` and expand/collapse.

## WCAG or Vendor Guidance (optional)

* [WCAG 2.0 No Keyboard Trap: Understanding SC 2.1.2](https://www.w3.org/TR/UNDERSTANDING-WCAG20/keyboard-operation-trapping.html) - If keyboard focus can be moved to a component of the page using a keyboard interface, then focus can be moved away from that component using only a keyboard interface, and, if it requires more than unmodified arrow or tab keys or other standard exit methods, the user is advised of the method for moving focus away. (Level A)

* [W3C - Ensuring that users are not trapped in content](https://www.w3.org/TR/WCAG20-TECHS/G21.html)

## Screenshots or Trace Logs

![Screen Shot 2020-07-08 at 7 59 28 AM](https://user-images.githubusercontent.com/57469/86916095-0f9fbe80-c0f1-11ea-8b79-811a4bfcc32f.png)

![Screen Shot 2020-07-08 at 7 59 51 AM](https://user-images.githubusercontent.com/57469/86916109-14647280-c0f1-11ea-8351-918b50a51150.png)

![Screen Shot 2020-07-08 at 8 00 05 AM](https://user-images.githubusercontent.com/57469/86916124-17f7f980-c0f1-11ea-9347-c43c80a25119.png)

### Examples of the accordion pattern

![Screen Shot 2020-07-08 at 7 45 40 AM](https://user-images.githubusercontent.com/57469/86916166-30681400-c0f1-11ea-843c-09c2dfe11b47.png)

![Screen Shot 2020-07-08 at 7 45 59 AM](https://user-images.githubusercontent.com/57469/86916179-34943180-c0f1-11ea-9417-e36b2e71c48c.png)

![Screen Shot 2020-07-08 at 7 46 11 AM](https://user-images.githubusercontent.com/57469/86916185-3827b880-c0f1-11ea-89f9-3a7d89d77f72.png)

**Comment by jenstrickland** [comment]
cc @1Copenut for review

**Comment by 1Copenut** [comment]
I agree with @jenstrickland on the keyboard trap and the fix. We should also make sure the focus halo appears on all four sides of the link or button that has current focus. The yellow halo is an important facet of inclusive design, and should not be cut off by padding or collapsed margins.

---
![Screen Shot 2020-07-08 at 9 10 53 AM](https://user-images.githubusercontent.com/934879/86929275-48449580-c0fb-11ea-8e7e-81b6ccb8e1f7.png)

**Comment by jenstrickland** [comment]
There's an open ticket documented for the focus halo issue — https://github.com/department-of-veterans-affairs/va.gov-team/issues/8975.

**Comment by davidconlon** [comment]
@mmiddaugh let's be sure to groom this one to see if we can pull it in to the next sprint.

@jenstrickland @1Copenut I'm a little concerned that we discovered a launchblocker on this when we spent a _significant_ amount of time held up in the Review Cycle on left nav.

**Comment by jenstrickland** [comment]
@davidconlon I share your concern. There are two things involved, in my assessment. 

1. My over-allocation is a risk I've communicated. We just received approval to hire another VSA accessibility specialist, which we asked for over the past six months. The hiring process begins next week.

1. VSP requires staging accessibility reviews be completed in two days. Staging reviews are less stringent — as documented — due to this. The Full Accessibility Audit & 508 Office Review is when we check more thoroughly. I only happened to find this issue while supporting the team addressing another issue (and it wouldn't generally be found in the staging review process). I pushed back on the 2-day requirement, stating quality would suffer, but was over-ruled. Additionally, the VSP Collab Cycle is still being iterated on, which adds to the complexity of reviews — ensuring requirements were complete, etc.

**Comment by mmiddaugh** [comment]
@jenstrickland this is ready to Validate when you are able

**Comment by jenstrickland** [comment]
This is not yet addressed. When the viewport is small and the In this section navigation collapses, I am still able to tab through the items within while the nav is collapsed.

**Comment by brangi** [comment]
@jenstrickland, what are the steps you use to reproduce this?

**Comment by brangi** [comment]
Thanks. However,  I am not able to reproduce it. Maybe the release was not deployed at that time. I will check tomorrow with @mpelzsherman.

**Comment by mpelzsherman** [comment]
I was also unable to reproduce the steps @jenstrickland documented above. Can you please try again?

**Comment by jenstrickland** [comment]
There must have been an issue with the release to staging. When I tabbed yesterday, I went through the items in the nav. I just tested it now and it went to the next interactive element outside of the nav, as it should.
